### ISSUE TITLE
[Bug] Save File dialog allows "saving" file in a folder not permitted but file never gets saved

### ISSUE THREAD

**Issue by hyuri** [flatpak/xdg-desktop-portal#1029]
Hi!

If an app has permission to ~/Documents, when I open the save file dialog, it lets me "save" the file in ~/, even though the file never gets saved. If I close the app, and go to ~/, the file I just "saved" isn't there.

I just lost a file I saved in ~/ because the app only had permission to ~/Documents but the portal still pretended to save it in ~/.

**Comment by A6GibKm** [comment]
What app is this? It shoulds like the app wasn't actually using the portal

**Comment by hyuri** [comment]
It happened to more than one app. But the one I recall now is [Story Architect](https://github.com/story-apps/starc).

On Flathub:
https://flathub.org/apps/dev.storyapps.starc

**Comment by smcv** [comment]
It is not technically possible for the sandbox to contain `/home/you/Documents` but not `/home/you` - there has to be *something* at `/home/you`. The current implementation is that the parent directory is an empty tmpfs.

Ideally it would be a read-only tmpfs, but that's more difficult to implement: `FlatpakExports` would have to instruct bubblewrap to create `/home/you` as a read/write tmpfs, then create the `Documents` subdirectory (and all other subdirectories), and then remount `/home/you` (and any other mount points below that point that are not meant to be available) as read-only. I'd be happy to review a merge request if someone wants to look into implementing that.

**Comment by smcv** [comment]
> It sounds like the app wasn't actually using the portal

Yes, apps that use the [file chooser portal](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.FileChooser) shouldn't have this issue.

**Comment by hyuri** [comment]
So is this really a portal / file save bug, or should I close this and report it to the individual apps instead?

**Comment by A6GibKm** [comment]
I personally don't have any experience with those apps, or the framework they are written, but your issue sounds exactly like they are not using the file chooser portal when they should be doing so. Please report an issue on their issue trackers.

Note that modern toolkits usually have helpers to use the file chooser portal.

**Comment by smcv** [comment]
> It is not technically possible for the sandbox to contain `/home/you/Documents` but not `/home/you` - there has to be _something_ at `/home/you`. The current implementation is that the parent directory is an empty tmpfs.
> 
> Ideally it would be a read-only tmpfs, but that's more difficult to implement: `FlatpakExports` would have to instruct bubblewrap to create `/home/you` as a read/write tmpfs, then create the `Documents` subdirectory (and all other subdirectories), and then remount `/home/you` (and any other mount points below that point that are not meant to be available) as read-only. I'd be happy to review a merge request if someone wants to look into implementing that.

Sorry, I thought this was a Flatpak bug report when I wrote that, but it's actually an xdg-desktop-portal bug report.

You would be able to see whether the app is using the file chooser portal or not by watching `dbus-monitor` output - the file chooser protocol involves quite a lot of D-Bus messages. For comparison and so that you can see what to expect, `org.gnome.Recipes` is an example of a small app that does use the file chooser portal (choose to create a new recipe and then browse for a photo).

If you are encouraged to add `--filesystem` permissions to a Flatpak app, that's often because it doesn't use the file chooser portal. If that's the case, then it is not using this project's functionality for that feature at all, so there is nothing that the `xdg-desktop-portal` project can do to change its behaviour.

https://github.com/flathub/dev.storyapps.starc/blob/master/dev.storyapps.starc.yml has `--filesystem=xdg-documents:rw`, which is a sign that it is probably not using the file chooser protocol (and also, a partial sandbox escape). Please open an issue asking for it to use the file chooser portal instead.

Similarly, https://github.com/flathub/org.onlyoffice.desktopeditors/blob/master/org.onlyoffice.desktopeditors.json has `--filesystem=host`, which is a sandbox escape. It seems to be an Electron app; if its maintainer upgrades to Electron 18, that should make it use the file chooser portal, which might mean that it no longer needs scary filesystem permissions (this was meant to be in Electron 14, but didn't actually work until it was accidentally fixed by a seemingly unrelated bug fix).

For apps where direct filesystem access is genuinely necessary, the Flatpak side of this (for which what I said above is still true) is https://github.com/flatpak/flatpak/issues/4813 and https://github.com/flatpak/flatpak/issues/1596.
